### ISSUE TITLE
How does one send a WebSocket message from server to client?

### ISSUE THREAD

**Issue by isiahmeadows** [ninenines/cowboy#1540]
The [docs](https://ninenines.eu/docs/en/cowboy/2.6/guide/ws_handlers/) include `self() ! <<"message to client">>` as an example somehow, but the [source](https://github.com/ninenines/cowboy/blob/master/src/cowboy_websocket.erl#L431-L433) [code](https://github.com/ninenines/cowboy/blob/master/src/cowboy_websocket.erl#L528) seems to indicate it's actually this:

```erl
send_request_somehow() ->
	self() ! <<"message to client">>.

websocket_info(Msg, State) when is_binary(Msg) ->
    {reply, {text, Msg}, State}.
```

Is my understanding of the source code correct, and could this be documented?

**Comment by bullno1** [comment]
https://ninenines.eu/docs/en/cowboy/2.6/guide/ws_handlers/#_receiving_erlang_messages: `websocket_info` is the callback that will be called by cowboy. Its return value **may** trigger a websocket frame being sent to the client.

The [next section](https://ninenines.eu/docs/en/cowboy/2.6/guide/ws_handlers/#_sending_frames) also goes into detail.

Sending a message to self is just **one** of the way to trigger `websocket_info`, that's why it is an example. In practice, what can send a message relies entirely on how you design your application.

So basically, it's already documented.

cowboy_websocket.erl abstracts all the implementation details away so your code only have to handle call back and return appropriate values.

I suggest trying to build something with `gen_server` first.
The concept is similar.

**Comment by isiahmeadows** [comment]
@bullno1 Thanks! To be fair, [`gen_server` has `gen_server:reply` which cuts the indirection](https://erlang.org/doc/man/gen_server.html#reply-2), hence why it threw me a little. I wish the combination of the two (`self() ! ...` -> return `{reply, ...}` from `websocket_info`) was better documented, though.

**Comment by essen** [comment]
Patch welcome!
